pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -168,13 +200,45 @@ func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode kubeletc
 
 // TODO: move thic logic into cni bridge plugin and remove this from kubenet
 func (plugin *kubenetNetworkPlugin) ensureMasqRule() error {
-	if plugin.nonMasqueradeCIDR != "0.0.0.0/0" {
-		if _, err := plugin.iptables.EnsureRule(utiliptables.Append, utiliptables.TableNAT, utiliptables.ChainPostrouting,
-			"-m", "comment", "--comment", "kubenet: SNAT for outbound traffic from cluster",
-			"-m", "addrtype", "!", "--dst-type", "LOCAL",
-			"!", "-d", plugin.nonMasqueradeCIDR,
-			"-j", "MASQUERADE"); err != nil {
-			return fmt.Errorf("Failed to ensure that %s chain %s jumps to MASQUERADE: %v", utiliptables.TableNAT, utiliptables.ChainPostrouting, err)
+	for _, currentCIDR := range plugin.nonMasqueradeCIDRs {
+		if currentCIDR != zeroCIDRv4 && currentCIDR != zeroCIDRv6 {
+			// switch according to target nonMasqueradeCidr ip family
+			ipt := plugin.iptables
+			if netutils.IsIPv6CIDRString(currentCIDR) {
+				ipt = plugin.iptablesv6
+			}
+
+			if _, err := ipt.EnsureRule(utiliptables.Append, utiliptables.TableNAT, utiliptables.ChainPostrouting,
+				"-m", "comment", "--comment", "kubenet: SNAT for outbound traffic from cluster",
+				"-m", "addrtype", "!", "--dst-type", "LOCAL",
+				"!", "-d", currentCIDR,
+				"-j", "MASQUERADE"); err != nil {
+				return fmt.Errorf("Failed to ensure that %s chain %s jumps to MASQUERADE: %v", utiliptables.TableNAT, utiliptables.ChainPostrouting, err)
+			}
+		}
+	}
+
+	//TODO: khenidak
+	// We are masquerading the secondary address for dual stack here
```

Comment:
As discussed on slack, I don't understand this block at all

Reply:
We have to masquerade the secondary IP. I can not change kube-proxy in the scope of the phase-1 (by adding concurrent v6,v4 proxiers), because it will even increase the size and complexity of the PR and accordingly the risk. So this is a stop gap until phase two is done. Phase 2 will modify kube-proxy to correctly snat all IPs assigned to pod. Unfortunately this has the side effect of s-natting the secondary IP even for pod on different nodes.